### PR TITLE
Feature/5275 clean progress bar print

### LINKED ISSUE #5275: Clean printing to terminal alongside progress bar
## 🚀 Feature
A way to print to terminal without breaking up the progress bar.

### Motivation
A lot of people print stuff to terminal while training/validating/testing, and currently a simple call to print() will break the progress bar. A way to get around this is to set up a custom progress bar, with methods for calling tqdm.write, and passing that as a callback to the trainer. However, this feels like a lot of effort for just getting clean terminal output alongside the progress bar.

### Pitch

Ability to get the ProgressBar - or the current active tqdm instance (main_progress_bar, val_progres_bar, etc) - through the Trainer or the LightningModule.
Something like pbar = trainer.pbar would feel intuitive.
Then, the user can call pbar.write(), and get clean printing like tqdm.write().

### PR DESCRIPTION
**Pull request by asnorkin**
## What does this PR do?

Fixes #5275

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [x] **Check that target branch and milestone match!**

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/core/lightning.py [hunk 1]
```diff
@@ -182,7 +182,11 @@ def forward(self, x):
 
         """
         if self.trainer.is_global_zero:
```
Comment (rohitgr7) [inline pytorch_lightning/core/lightning.py:184]: maybe add an argument here `prog_bar=True/False` just like we have for `self.log`? WDYT?

**Review by awaelchli — COMMENTED** [review]
Looking very good, could you also add a stub to the base class (ProgressBarBase)? Thanks

File: pytorch_lightning/core/lightning.py (hunk 1 above)
Comment (asnorkin) [reply pytorch_lightning/core/lightning.py:181]: I suppose that these are different cases.

For self.log prog_bar=True/False works as follows:

prog_bar=True
```self.log('my_metric', 0.427, prog_bar=True)```
Epoch 36:   9%|███▉                                      | 67/716 [01:43<16:46,  1.55s/it, loss=0.462, v_num=2, my_metric=0.427]

prog_bar=False
```self.log('my_metric', 0.427, prog_bar=False)```
Epoch 36:   9%|███▉                                      | 67/716 [01:43<16:46,  1.55s/it, loss=0.462, v_num=2]

For self.print we want not break progress bar. If we have ```self.print``` call in training step:
```
def training_step(self, batch, batch_idx):
    ...
    self.print('Hello world!')
    ...
```
It will give the following log:
...
Epoch 36:   9%|███▉                                      | 67/716 [01:43<16:46,  1.55s/it, loss=0.462, v_num=2]
Hello world!
Epoch 36:   9%|███▉                                      | 68/716 [01:45<16:44,  1.54s/it, loss=0.463, v_num=2]
Hello world!
Epoch 36:   9%|███▉                                      | 69/716 [01:46<16:43,  1.55s/it, loss=0.461, v_num=2]
...

But we want the following:
...
Hello world!
Hello world!
Epoch 36:   9%|███▉                                      | 689/716 [01:46<16:43,  1.55s/it, loss=0.461, v_num=2]


May be we can give user the choice of whether to break progress bar or not, but prog_bar=True will be confusing because it has different meaning than for self.log.

Comment (rohitgr7) [reply pytorch_lightning/core/lightning.py:181]: yep, it might be, maybe some other name? `use_prog_bar` or something.. Just a suggestion. Either way is fine.

Comment (rohitgr7) [reply pytorch_lightning/core/lightning.py:181]: ok I thought it writes in the progress bar as a description or something. looks good :+1:

**Review by rohitgr7 — APPROVED** [review]
LGTM apart from Adrian's comment.

File: docs/source/multi_gpu.rst [hunk 2]
```diff
@@ -654,7 +654,7 @@ To use Sharded Training, you need to first install FairScale using the command b
 
 .. code-block:: bash
 
-    pip install https://github.com/PyTorchLightning/fairscale/archive/pl_1.1.0.zip
+    pip install fairscale
```
Comment (carmocca) [inline docs/source/multi_gpu.rst:657]: the changes to this file should not be here

File: tests/metrics/test_metric.py [hunk 3]
```diff
@@ -40,6 +40,20 @@ def compute(self):
         pass
 
 
+class DummyList(Metric):
```
Comment (carmocca) [inline tests/metrics/test_metric.py:43]: what's this for?

**Review by carmocca — COMMENTED** [review]
Cool! Can we add tests?

File: docs/source/multi_gpu.rst (hunk 2 above)
Comment (rohitgr7) [reply docs/source/multi_gpu.rst:657]: looks like rebase wasn't done correctly.

File: pytorch_lightning/callbacks/progress.py [hunk 4]
```diff
@@ -373,6 +379,20 @@ def on_test_end(self, trainer, pl_module):
         super().on_test_end(trainer, pl_module)
         self.test_progress_bar.close()
 
+    def print(self, *args, sep=' ', end='\n', file=None, nolock=False):
```
Comment (Borda) [inline pytorch_lightning/callbacks/progress.py:382]:
```suggestion
    def print(self, *args, sep=' ', end: str = os.linesep, file=None, nolock=False):
```

**Review by Borda — COMMENTED** [review]
@asnorkin mind check comments from @awaelchli then I think we are ready to go... 🐰

**Review by SeanNaren — APPROVED** [review]

File: CHANGELOG.md [hunk 5]
```diff
@@ -118,6 +118,10 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Added `Trainer` flag to activate Stochastic Weight Averaging (SWA) `Trainer(stochastic_weight_avg=True)` ([#6038](https://github.com/PyTorchLightning/pytorch-lightning/pull/6038))
 
 
+- Added a way to print to terminal without breaking up the progress bar ([#5470](https://github.com/PyTorchLightning/pytorch-lightning/pull/5470))
+
+
+
```
Comment (carmocca) [inline CHANGELOG.md:124]:
```suggestion
```

**Review by carmocca — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !
